modules/python/python-http-header.c

```diff
@@ -0,0 +1,383 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "python-http-header.h"
+#include "python-helpers.h"
+
+#include "driver.h"
+#include "str-utils.h"
+#include "list-adt.h"
+
+#include "modules/http/http-signals.h"
+
+#include <time.h>
+
+#define PYTHON_HTTP_HEADER_PLUGIN "python-http-header"
+
+struct _PythonHttpHeaderPlugin
+{
+  LogDriverPlugin super;
+
+  gchar *class;
+  GList *loaders;
+
+  GList *last_headers;
+  time_t last_run;
+  int timeout;
+
+  GHashTable *options;
+
+  struct
+  {
+    PyObject *class;
+    PyObject *instance;
+    PyObject *get_headers;
+  } py;
+};
+
+static gboolean
+_py_append_pylist_to_list(PyObject *py_list, GList **list)
+{
+  const gchar *str;
+  PyObject *py_str;
+
+  if (!py_list || !PyList_Check(py_list))
+    goto exit;
+
+  Py_ssize_t len = PyList_Size(py_list);
+  for (int i = 0; i < len; i++)
+    {
+      py_str = PyList_GetItem(py_list, i); // Borrowed reference
+      if (!py_str || !PyUnicode_Check(py_str))
+        goto exit;
+
+      if (!(str = _py_get_string_as_string(py_str)))
+        goto exit;
+
+      *list = g_list_append(*list, g_strdup(str));
+    }
+
+  return TRUE;
+
+exit:
+  return FALSE;
+}
+
+static void
+_py_append_str_to_pylist(gconstpointer data, gpointer user_data)
+{
+  PyObject *py_str = PyUnicode_FromString((gchar *) data);
+  if (!py_str)
+    {
+      gchar buf[256];
+
+      msg_error("Error creating Python String object from C string",
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  PyObject *py_list = (PyObject *) user_data;
+  if (PyList_Append(py_list, py_str) != 0)
+    {
+      gchar buf[256];
+
+      msg_error("Error adding new item to Python List",
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+    }
+
+exit:
+  Py_XDECREF(py_str); // Even if append() succeeds, it doesn't steal the reference to py_str.
+}
+
+static PyObject *
+_py_convert_list_to_pylist(List *list)
+{
+  PyObject *py_list = PyList_New(0);
+  g_assert(py_list);
+
+  if (list)
+    list_foreach(list, _py_append_str_to_pylist, py_list);
+
+  return py_list;
+}
+
+static gboolean
+_py_attach_bindings(PythonHttpHeaderPlugin *self)
+{
+  PyObject *py_args = NULL;
+
+  self->py.class = _py_resolve_qualified_name(self->class);
```

Comment:
This may be a yagni. But ...

In case of python-source and fetcher, and in java binding, it is nice to have a superclass, that users need to inherit from. Through that class, we could inject data (members, methods, constants) cleaner. For now there is no need to inject such thing, but in the future we may need.

What we might consider is to have (and force users to use) a superclass, like `SyslogNGHttpBase`, even if it is a dummy class for now (maybe put into `modules/python/pylib/syslogng/__init__.py`, and then we might avoid writing c, and implement such things in python, similarly to https://github.com/syslog-ng/syslog-ng/pull/3100/commits/5f6ac2cdcfc724e1a043845ff5f78d2510a9d5de.)

Later it would be difficult to introduce a superclass and maintain backward compatibility at the same time.

Reply:
I think that signal-slot interfaces should be light and slots should not depend on modules who emits the signals they are connected to.

The general python http header is a slot implementation and the interface between this module and http module is the signal-slot interface.